src/components/lib/vmrt/vmrt.c

```diff
@@ -0,0 +1,427 @@
+{
+	struct vm_vcpu_shared_region *regs = vcpu->shared_region;
+
+	printc("\tRAX: %016llx\tRBX: %016llx\tRCX: %016llx\tRDX: %016llx\n",
+		regs->ax, regs->bx, regs->cx, regs->dx);
+	printc("\tR8:  %016llx\tR9:  %016llx\tR10: %016llx\tR11: %016llx\n",
+		regs->r8, regs->r9, regs->r10, regs->r11);
+	printc("\tR12: %016llx\tR13: %016llx\tR14: %016llx\tR15: %016llx\n",
+		regs->r12, regs->r13, regs->r14, regs->r15);
+	printc("\tRSI: %016llx\tRDI: %016llx\tRBP: %016llx\n",
+		regs->si, regs->di, regs->bp);
+
+	printc("\tRIP: %016llx\tRSP: %016llx\tInst Len: %llu\n", regs->ip, regs->sp, regs->inst_length);
+	printc("\tGPA: %016llx\n", regs->gpa);
+	printc("\tEFER: %016llx\t\n", regs->efer);
+	printc("\tCR0: %016llx\tCR2: %016llx\tCR4: %016llx\n", regs->cr0, regs->cr2, regs->cr4);
+}
+
+void
+vmrt_vmcs_page_create(struct vmrt_vm_vcpu *vcpu)
+{
+	vcpu->vmcs_cap = capmgr_vm_vmcs_create();
+}
+
+void
+vmrt_msr_bitmap_page_create(struct vmrt_vm_vcpu *vcpu)
+{
+	vcpu->msr_bitmap_cap = capmgr_vm_msr_bitmap_create();
+}
+
+void
+vmrt_lapic_page_create(struct vmrt_vm_vcpu *vcpu)
+{
+	vaddr_t page;
+
+	vcpu->lapic_cap = capmgr_vm_lapic_create(&page);
+	vcpu->lapic_page = (void *)page;
+}
+
+void
+vmrt_shared_page_create(struct vmrt_vm_vcpu *vcpu)
+{
+	vaddr_t page;
+
+	vcpu->shared_mem_cap = capmgr_vm_shared_region_create(&page);
+	vcpu->shared_region = (struct vm_vcpu_shared_region *)page;
+}
+
+void
+vmrt_lapic_access_page_create(struct vmrt_vm_vcpu *vcpu)
+{
+	vcpu->lapic_access_cap = capmgr_vm_lapic_access_create(vcpu->vm->lapic_access_page);
+}
+
+void
+vmrt_vmcb_create(struct vmrt_vm_vcpu *vcpu)
+{
+	vcpu->vmcb_cap = capmgr_vm_vmcb_create(vcpu->vmcs_cap, vcpu->msr_bitmap_cap, vcpu->lapic_access_cap, vcpu->lapic_cap, vcpu->shared_mem_cap, vcpu->handler_tid << 16| vcpu->vpid, 0);
+}
+
+void
+vmrt_vm_vcpu_init(struct vmrt_vm_comp *vm, u32_t vcpu_nr)
+{
+	struct vmrt_vm_vcpu *vcpu;
+	thdid_t tid, handler_tid;
+	thdcap_t cap;
+	vaddr_t resource;
+	assert(vcpu_nr < vm->num_vpu);
+
+	vcpu = &vm->vcpus[vcpu_nr];
+	assert(!vcpu->shared_region);
+
+	vcpu->vm = vm;
+
+	handler_tid = sched_thd_create(vmrt_vm_exception_handler, vcpu);
+	vcpu->handler_tid = handler_tid;
+
+	vcpu->vpid = vcpu_nr + 1;
+	vmrt_vmcs_page_create(vcpu);
+	vmrt_msr_bitmap_page_create(vcpu);
+	vmrt_lapic_page_create(vcpu);
+	vmrt_shared_page_create(vcpu);
+	vmrt_lapic_access_page_create(vcpu);
+	vmrt_vmcb_create(vcpu);
+
+	cap = capmgr_vm_vcpu_create(vcpu->vm->comp_id, vcpu->vmcb_cap, &tid);
+	assert(tid != 0);
+	vcpu->tid = tid;
+	vcpu->cap = cap;
+	vcpu->cpuid = vcpu_nr;
+	vcpu->coreid = cos_coreid();
+
+	vcpu->next_timer = ~0ULL;
+
+}
+
+void
+vmrt_vm_mem_init(struct vmrt_vm_comp *vm, void *mem)
+{
+	vm->guest_addr = mem;
+
+	/* Clean the memory for the VM */
+	memset(mem, 0, vm->guest_mem_sz);
+}
+
+void
+vmrt_vm_create(struct vmrt_vm_comp *vm, char *name, u8_t num_vcpu, u64_t vm_mem_sz)
+{
+	compid_t vm_id = 0;
+	
+	vaddr_t lapic_access_page;
+
+	assert((vm_mem_sz % PAGE_SIZE_4K) == 0);
```

Comment:
I haven't commented on this yet, but I don't really understand the point of PAGE_SIZE_4K. Shouldn't it just be PAGE_SIZE? Might be missing something.

Reply:
Since we are using Intel's vmx support, and it's current EPT setting requires exactly 4K size page. However, the PAGE_SIZE definition can be different based on different arch and page table settings(huge page). Thus, I believe we need to be more specific here.